### PR TITLE
Fixes #158: Add support for Terraform v0.12

### LINKED ISSUE #158: Add support for Terraform v0.12
Terraform v0.12.0
+ provider.google v2.7.0
+ provider.google-beta v2.7.0
+ provider.kubernetes v1.7.0
+ provider.null v2.1.2
+ provider.random v2.1.2

Initializing modules...
- gke in ../..

Initializing the backend...

Initializing provider plugins...
- Checking for available provider plugins...
- Downloading plugin for provider "google-beta" (terraform-providers/google-beta) 2.7.0...
- Downloading plugin for provider "null" (terraform-providers/null) 2.1.2...
- Downloading plugin for provider "random" (terraform-providers/random) 2.1.2...
- Downloading plugin for provider "kubernetes" (terraform-providers/kubernetes) 1.7.0...
- Downloading plugin for provider "google" (terraform-providers/google) 2.7.0...

The following providers do not have any version constraints in configuration,
so the latest version was installed.

To prevent automatic upgrades to new major versions that may contain breaking
changes, it is recommended to add version = "..." constraints to the
corresponding provider blocks in configuration, with the constraint strings
suggested below.

* provider.kubernetes: version = "~> 1.7"
* provider.null: version = "~> 2.1"
* provider.random: version = "~> 2.1"

Terraform has been successfully initialized!

You may now begin working with Terraform. Try running "terraform plan" to see
any changes that are required for your infrastructure. All Terraform commands
should now work.

If you ever set or change modules or backend configuration for Terraform,
rerun this command to reinitialize your working directory. If you forget, other
commands will detect it and remind you to do so if necessary.
root@gcloud-gke-ubuntu:/home/vagrant/terraform-google-kubernetes-engine/examples/simple_zonal# clear
root@gcloud-gke-ubuntu:/home/vagrant/terraform-google-kubernetes-engine/examples/simple_zonal# terraform plan
var.ip_range_pods
  The secondary ip range to use for pods

  Enter a value:

var.ip_range_pods
  The secondary ip range to use for pods

  Enter a value:

var.ip_range_pods
  The secondary ip range to use for pods

  Enter a value:

var.ip_range_pods
  The secondary ip range to use for pods

  Enter a value: 10.8.0.0

var.ip_range_services
  The secondary ip range to use for pods

  Enter a value: 10.8.0.0

var.network
  The VPC network to host the cluster in

  Enter a value: default

var.project_id
  The project ID to host the cluster in

  Enter a value: mystical-being-195215

var.region
  The region to host the cluster in

  Enter a value: us-central1

var.subnetwork
  The subnetwork to host the cluster in

  Enter a value: default

var.zones
  The zone to host the cluster in (required if is a zonal cluster)

  Enter a value: us-central1-a

Error: Variables not allowed

  on <value for var.zones> line 1:
  (source code not available)

Variables may not be used here.

root@gcloud-gke-ubuntu:/home/vagrant/terraform-google-kubernetes-engine/examples/simple_zonal# terraform plan
var.ip_range_pods
  The secondary ip range to use for pods

  Enter a value: 10.8.0.0

var.ip_range_services
  The secondary ip range to use for pods

  Enter a value: 10.8.0.0

var.network
  The VPC network to host the cluster in

  Enter a value: default

var.project_id
  The project ID to host the cluster in

  Enter a value:
var.ip_range_pods

var.project_id
  The project ID to host the cluster in

  Enter a value:
var.region
  The secondary ip range to use for pods
  The region to host the cluster in

  Enter a value:
var.subnetwork
  The subnetwork to host the cluster in

  Enter a value:

var.subnetwork
  The subnetwork to host the cluster in

  Enter a value: mystical-being-195215

var.zones
  The zone to host the cluster in (required if is a zonal cluster)

  Enter a value: ["us-central1-a", "us-central1-b", "us-central1-f"]

Error: Unsupported argument

  on ../../cluster_regional.tf line 39, in resource "google_container_cluster" "primary":
  39:   master_authorized_networks_config = ["${var.master_authorized_networks_config}"]

An argument named "master_authorized_networks_config" is not expected here.
Did you mean to define a block of type "master_authorized_networks_config"?

Error: Incorrect attribute value type

  on ../../cluster_regional.tf line 52, in resource "google_container_cluster" "primary":
  52:       disabled = "${var.http_load_balancing ? 0 : 1}"

Inappropriate value for attribute "disabled": bool required.

Error: Incorrect attribute value type

  on ../../cluster_regional.tf line 56, in resource "google_container_cluster" "primary":
  56:       disabled = "${var.horizontal_pod_autoscaling ? 0 : 1}"

Inappropriate value for attribute "disabled": bool required.

Error: Incorrect attribute value type

  on ../../cluster_regional.tf line 60, in resource "google_container_cluster" "primary":
  60:       disabled = "${var.kubernetes_dashboard ? 0 : 1}"

Inappropriate value for attribute "disabled": bool required.

Error: Incorrect attribute value type

  on ../../cluster_regional.tf line 64, in resource "google_container_cluster" "primary":
  64:       disabled = "${var.network_policy ? 0 : 1}"

Inappropriate value for attribute "disabled": bool required.

Error: Incorrect attribute value type

  on ../../cluster_zonal.tf line 30, in resource "google_container_cluster" "zonal_primary":
  30:   node_locations = ["${slice(var.zones,1,length(var.zones))}"]

Inappropriate value for attribute "node_locations": element 0: string
required.

Error: Unsupported argument

  on ../../cluster_zonal.tf line 39, in resource "google_container_cluster" "zonal_primary":
  39:   master_authorized_networks_config = ["${var.master_authorized_networks_config}"]

An argument named "master_authorized_networks_config" is not expected here.
Did you mean to define a block of type "master_authorized_networks_config"?

Error: Incorrect attribute value type

  on ../../cluster_zonal.tf line 52, in resource "google_container_cluster" "zonal_primary":
  52:       disabled = "${var.http_load_balancing ? 0 : 1}"

Inappropriate value for attribute "disabled": bool required.

Error: Incorrect attribute value type

  on ../../cluster_zonal.tf line 56, in resource "google_container_cluster" "zonal_primary":
  56:       disabled = "${var.horizontal_pod_autoscaling ? 0 : 1}"

Inappropriate value for attribute "disabled": bool required.

Error: Incorrect attribute value type

  on ../../cluster_zonal.tf line 60, in resource "google_container_cluster" "zonal_primary":
  60:       disabled = "${var.kubernetes_dashboard ? 0 : 1}"

Inappropriate value for attribute "disabled": bool required.

Error: Incorrect attribute value type

  on ../../cluster_zonal.tf line 64, in resource "google_container_cluster" "zonal_primary":
  64:       disabled = "${var.network_policy ? 0 : 1}"

Inappropriate value for attribute "disabled": bool required.

Error: Missing resource instance key

  on ../../sa.tf line 40, in resource "google_project_iam_member" "cluster_service_account-log_writer":
  40:   project = "${google_service_account.cluster_service_account.project}"

Because google_service_account.cluster_service_account has "count" set, its
attributes must be accessed on specific instances.

For example, to correlate with indices of a referring resource, use:
    google_service_account.cluster_service_account[count.index]

Error: Missing resource instance key

  on ../../sa.tf line 42, in resource "google_project_iam_member" "cluster_service_account-log_writer":
  42:   member  = "serviceAccount:${google_service_account.cluster_service_account.email}"

Because google_service_account.cluster_service_account has "count" set, its
attributes must be accessed on specific instances.

For example, to correlate with indices of a referring resource, use:
    google_service_account.cluster_service_account[count.index]

### PR DESCRIPTION
**Pull request by bohdanyurov-gl**
https://github.com/terraform-google-modules/terraform-google-kubernetes-engine/issues/158

Add support for TF 0.12.
- autogen
- root and private_cluster modules
- tests (including support for validation)
- examples

### CODE REVIEW

**Comment by aaron-lane** [comment]
Fixes #158.

File: test/make.sh [hunk 1]
```diff
@@ -45,11 +45,20 @@ function docker() {
...
+    -not -path "./test/fixtures/shared/*" \
+    -not -path "./test/fixtures/all_examples/*" \
+    | xargs dirname | sort | uniq | xargs -L 1 -i{} bash -c 'terraform init "{}" > /dev/null && terraform validate "{}"'
+
   echo "Running terraform fmt"
   #shellcheck disable=SC2156
-  find . -name "*.tf" -not -path "./autogen/*" -not -path "./test/fixtures/shared/*" -not -path "./test/fixtures/all_examples/*" -exec bash -c 'terraform fmt -check=true -write=false "{}"' \;
+  find . -name "*.tf" \
+    -not -path "./autogen/*" \
+    -not -path "./test/fixtures/shared/*" \
```
Comment (aaron-lane) [inline test/make.sh]: Shared files should be formatted.
```suggestion
```

File: test/make.sh [hunk 2]
```diff
@@ -45,11 +45,20 @@ function docker() {
...
+    -not -path "./test/fixtures/all_examples/*" \
```
Comment (aaron-lane) [inline test/make.sh]: Shared files should be formatted.
```suggestion
```

File: test/make.sh [hunk 3]
```diff
@@ -45,11 +45,20 @@ function docker() {
...
+    -exec bash -c 'terraform fmt -check=true -write=false "{}"' \;
```
Comment (aaron-lane) [inline test/make.sh]: `terraform fmt` should only be run for each directory.

```suggestion
| xargs dirname | sort | uniq | xargs -L 1 -i{} bash -c 'terraform fmt "{}"'
```

File: test/make.sh (hunk 1 above)
Comment (bohdanyurov-gl) [reply test/make.sh]: Isn't it out of scope? I haven't changed it

File: autogen/cluster_regional.tf [hunk 4]
```diff
@@ -20,74 +20,90 @@
...
   network_policy {
-    enabled  = "${var.network_policy}"
-    provider = "${var.network_policy_provider}"
+    enabled  = var.network_policy
+    provider = var.network_policy_provider
   }
 
-  subnetwork         = "${replace(data.google_compute_subnetwork.gke_subnetwork.self_link, "https://www.googleapis.com/compute/v1/", "")}"
-  min_master_version = "${local.kubernetes_version_regional}"
+  subnetwork = data.google_compute_subnetwork.gke_subnetwork.self_link
```
[the remark this answers is not in the document]
Comment (bohdanyurov-gl) [reply autogen/cluster_regional.tf]: Related to this issue:
https://github.com/terraform-google-modules/terraform-google-kubernetes-engine/issues/75

File: test/make.sh (hunk 1 above)
Comment (aaron-lane) [reply test/make.sh]: Sorry, you're right.

**Comment by aaron-lane** [comment]
Fixes #75.

**Pull request merged**